File: leetcode-js/114-flatten-binary-tree-to-linked-list.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) {
    helper(root, null);
};

function helper(root, prev) {
    if (!root) {
        return prev;
    }

    prev = helper(root.right, prev);
    prev = helper(root.left, prev);

    root.right = prev;
    root.left = null;

    return root;
}

File: leetcode-java/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        helper(root, null);
    }

    TreeNode helper(TreeNode node, TreeNode prev) {
        if (node == null) {
            return prev;
        }

        prev = helper(node.right, prev);
        prev = helper(node.left, prev);

        node.right = prev;
        node.left = null;

        return node;
    }
}
